File: C++/cpp_lab_2/rational.cpp
#include "rational.h"
#include <cstdlib>

// Complete these methods:

int rational::gcd( int n1, int n2 )
{
  if (n2 == 0)
      return n1;
  return rational::gcd( n2, n1 % n2 );
}

void rational::normalize( )
{

  if (denum == 0)
      throw "Cannot divide by 0";

  int factor = rational::gcd(num, denum);
  num = num / factor;
  denum = denum / factor;

  if (denum < 0)
  {
      num = -num;
      denum = -denum;
  }
}

rational operator - ( rational r )
{
  return rational( -r.num, r.denum);
}

rational operator + ( const rational& r1, const rational& r2 )
{
  return rational( r1.num * r2.denum + r2.num * r1.denum, r1.denum * r2.denum);
}

rational operator - ( const rational& r1, const rational& r2 )
{
  return rational( r1.num * r2.denum - r2.num * r1.denum, r1.denum * r2.denum);
}

rational operator * ( const rational& r1, const rational& r2 )
{
  return rational( r1.num * r2.num, r1.denum * r2.denum );
}

rational operator / ( const rational& r1, const rational& r2 )
{
  return rational( r1.num * r2.denum, r2.num * r1.denum);
}

bool operator == ( const rational& r1, const rational& r2 )
{
  return r1.num == r2.num && r1.denum == r2.denum;
}

bool operator != ( const rational& r1, const rational& r2 )
{
  return r1.num != r2.num || r1.num != r2.num;
}

std::ostream& operator << ( std::ostream& out, const rational& r )
{
  if (r.denum == 1)
      out << r.num;
  else
      out << r.num << "/" << r.denum;
  return out;
}

File: Python/python_lab_3/rational.py
from numbers import *

def gcd(n1, n2) :
    if n1 == 0 and n2 == 0: raise ArithmeticError( "gcd(0,0) does not exist" )
    if n2 == 0: return abs(n1)
    return gcd(n2, n1 % n2)

class Rational( Number ) :
    def __init__( self, num, denum = 1 ) :
        self.num = num
        self.denum = denum
        self.normalize( )

    def normalize(self) :
        if self.denum == 0:
            raise ZeroDivisionError( "you cannot divide by 0" )
        if self.denum < 0:
            self.num = -self.num
            self.denum = -self.denum

        factor = gcd(self.num, self.denum)
        self.num = self.num // factor
        self.denum = self.denum // factor

    def __repr__(self) :
        if self.denum == 1:
            return "{}".format(self.num)
        else:
            return "{}/{}".format(self.num, self.denum)

    def __neg__( self ) :
        return Rational( -self.num, self.denum )

    def __add__( self, other ) :
        if not isinstance(other, Rational):
            return Rational( self.num + self.denum * other, self.denum )
        else:
            return Rational( self.num * other.denum + other.num * self.denum, self.denum * other.denum )

    def __radd__( self, other ) :
        return self.__add__( other )

    def __sub__( self, other ) :
        if not isinstance(other, Rational):
            return Rational( self.num - self.denum * other, self.denum )
        else:
            return Rational( self.num * other.denum - other.num * self.denum, self.denum * other.denum )

    def __rsub__( self, other ) :
        if not isinstance(other, Rational):
            return Rational( other * self.denum - self.num, self.denum )
        else:
            return Rational( other.num * self.denum - self.num * other.denum, self.denum * other.denum )

    def __mul__( self, other ) :
        if not isinstance(other, Rational):
            return Rational( self.num * other, self.denum )
        else:
            return Rational( self.num * other.num, self.denum * other.denum )

    def __rmul__( self, other ) :
        return self.__mul__(other)

    def __truediv__( self, other ) :
        if not isinstance(other, Rational):
            return Rational( self.num, self.denum * other )
        else:
            return Rational( self.num * other.denum, self.denum * other.num )

    def __rtruediv__( self, other ) :
        if not isinstance(other, Rational):
            return Rational( other * self.denum, self.num )
        else:
            return Rational( other.num * self.denum, self.num * other.denum )
